src/main/java/com/checkmarx/intellij/service/AuthService.java

```diff
@@ -0,0 +1,387 @@
+                    throw new RuntimeException(e); // wrap checked in unchecked to satisfy Supplier
+                }
+            }, MAX_RETRIES, RETRY_DELAY_MS);
+        } catch (Exception exception) {
+            LOGGER.error("OAuth: Failed to get OAuth token after retries. Root Cause:{} ", exception.getMessage());
+            return Collections.emptyMap();
+        }
+    }
+
+
+    /**
+     * This method will be used to call the token endpoint to get the OAuth token
+     * by using the auth code received from the authorization endpoint.
+     *
+     * @param code         - auth code received from authorization
+     * @param codeVerifier - original code verifier for PKCE
+     * @param redirectUri  - redirect uri, which was used for authorization code
+     * @return string refresh token
+     * @throws IOException - if error occurred during exchanging auth code for token
+     */
+    public Map<String, Object> callTokenEndpoint(String tokenEndpoint, String code, String codeVerifier, String redirectUri) throws IOException, InterruptedException {
+        int redirectAttempt = 1;
+
+        String requestBody = "grant_type=authorization_code" +
+                "&code=" + code +
+                "&redirect_uri=" + redirectUri +
+                "&client_id=" + Constants.AuthConstants.OAUTH_IDE_CLIENT_ID +
+                "&code_verifier=" + codeVerifier;
+        HttpClient client = HttpClient.newHttpClient();
+        do {
+            HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create(tokenEndpoint))
+                    .header("Content-Type", "application/x-www-form-urlencoded")
+                    .POST(HttpRequest.BodyPublishers.ofString(requestBody))
+                    .build();
+
+            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+
+            if (response != null) {
+                if (response.statusCode() == 200) {
+                    return extractRefreshTokenDetails(response.body());
+                } else if (response.statusCode() == 308) {
+                    LOGGER.warn(String.format("OAuth: Received permanent redirect, getting new token endpoint from header location. Attempt:%d", redirectAttempt));
+                    /*
+                     * If user provided only base url then, CxOne send permanent redirect response.
+                     * As token base url should be iam, we need to get new token endpoint from the
+                     * received response header and use this new endpoint to get the token.
+                     */
+                    if (response.headers() != null && (response.headers().map() != null && !response.headers().map().isEmpty())
+                            && response.headers().map().containsKey(Constants.AuthConstants.LOCATION)) {
+                        tokenEndpoint = response.headers().map().get(Constants.AuthConstants.LOCATION).get(0);
+                    }
+                }
```

Comment:
add 'else' to print a log for other status codes for which code will do another retry